[Merged by Bors] - refactor(Algebra/Category/ModuleCat): make `ModuleCat.Hom` a structure

Vierkantor opened the pull request:
This PR follows the approach of #19065 to solve painful definitional equalities problems by forbidding the identification of `A ⟶ B` and `A →ₗ[R] B`. These are not equal reducibly, so tactics get confused if one is replaced by the other.

There seem to be a few regressions caused by defeq abuse that we were able to get away with previously, but some subtle elaboration differences made it evident. Otherwise the cost is inserting many `.hom`s and `asHom`s.

A few steps that might clean up the diff, but would be too much work to incorporate in this PR:
 * Make the `ext` tactic pick up `TensorProduct.ext` again.
 * Replace the defeq abuse between `(restrictScalars f).obj M` and `M` with explicit maps going back and forth.

Overall quite a few proofs could be cleaned up at the cost of more bookkeeping.

---

- [x] depends on: #19705

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
refactor(Algebra/Category/ModuleCat): make `ModuleCat.Hom` a structure

This PR follows the approach of #19065 to solve painful definitional equalities problems by forbidding the identification of `A ⟶ B` and `A →ₗ[R] B`. These are not equal reducibly, so tactics get confused if one is replaced by the other.

Currently the PR consists mostly of me playing "type tetris" to get everything to compile again: I haven't looked for which proofs can be cleaned up as a result.

This commit is not quite finished, please see the Zulip thread for more details.

Vierkantor commented on Mathlib/Algebra/Category/ModuleCat/ChangeOfRings.lean:
```diff
@@ -76,24 +77,19 @@ def restrictScalars {R : Type u₁} {S : Type u₂} [Ring R] [Ring S] (f : R →
-  inferInstanceAs <| Module R <| RestrictScalars.obj' f M
-
--- Porting note: this should be automatic
+-- TODO: this instance gives diamonds if `f : S →+* S`...
```
This instance causes downstream issues in `Mathlib/Algebra/Category/ModuleCat/Presheaf/Pushforward.lean:40`, where the synthesized instance does not equal the one obtained by unification.

Vierkantor commented on Mathlib/Algebra/Category/ModuleCat/ChangeOfRings.lean:
```diff
@@ -129,14 +125,14 @@ morphisms `M ⟶ (ModuleCat.restrictScalars f).obj N`. -/
     (M : ModuleCat.{v} R) (N : ModuleCat.{v} S) :
     (M →ₛₗ[f] N) ≃+ (M ⟶ (ModuleCat.restrictScalars f).obj N) where
-  toFun g :=
+  toFun g := asHom (X := M) (Y := (ModuleCat.restrictScalars f).obj N) <|
```
Having to type-ascribe `asHom` suggests there is some defeq abuse going on.

Vierkantor commented on Mathlib/Algebra/Category/ModuleCat/ChangeOfRings.lean:
```diff
@@ -445,51 +434,42 @@ variable {R : Type u₁} {S : Type u₂} [Ring R] [Ring S] (f : R →+* S)
-      dsimp
-      rw [mul_smul]
+    (g : (restrictScalars f).obj Y ⟶ X) : Y ⟶ (coextendScalars f).obj X :=
+  asHom
```
The `@[simps]` lemma generated here has the wrong type, but if we type-ascript `asHom` to give it the right type, then the proofs in this definition breaks.

Vierkantor commented on Mathlib/Algebra/Category/ModuleCat/ChangeOfRings.lean:
```diff
@@ -569,26 +546,30 @@ def restrictCoextendScalarsAdj {R : Type u₁} {S : Type u₂} [Ring R] [Ring S]
-        right_inv := fun g => LinearMap.ext fun x => LinearMap.ext fun s : S => by
+          rw [RestrictionCoextensionAdj.HomEquiv.toRestriction_hom_apply, AddHom.toFun_eq_coe,
+            LinearMap.coe_toAddHom]
+          -- This needs to be `erw` because `coextendScalars.obj Y` is unfolded in
+          -- `fromRestriction_hom_apply_apply`. Changing `fromRestriction` allows the whole proof
+          -- to be `simp` but causes defeq trouble in the definition itself...)
```
More signs of defeq abuse.

Vierkantor commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -174,69 +242,42 @@ def ofSelfIso (M : ModuleCat R) : ModuleCat.of R M ≅ M where
-def ModuleCat.asHomRight [AddCommGroup X₁] [Module R X₁] {X₂ : ModuleCat.{v} R} :
-    (X₁ →ₗ[R] X₂) → (ModuleCat.of R X₁ ⟶ X₂) :=
-  id
+def ModuleCat.asHomRight [AddCommGroup X₁] [Module R X₁] {X₂ : ModuleCat.{v} R}
```
This should be defeq now to `asHom`. Can we safely replace it? (i.e. deprecated alias it)

Vierkantor commented on Mathlib/Algebra/Category/ModuleCat/Monoidal/Closed.lean:
```diff
@@ -21,20 +21,28 @@ namespace ModuleCat
 
 variable {R : Type u} [CommRing R]
 
+-- Not sure how to define these, presumably there's some abstract machine that does exactly this!
+def asHom₂ {M N P : ModuleCat.{u} R} (f : M →ₗ[R] N →ₗ[R] P) :
+    M ⟶ ((linearCoyoneda R (ModuleCat R)).obj (op N)).obj P := sorry
+
+-- Not sure how to define these, presumably there's some abstract machine that does exactly this!
+def Hom.hom₂ {M N P : ModuleCat.{u} R}
+    (f : ModuleCat.Hom M (((linearCoyoneda R (ModuleCat R)).obj (op N)).obj P)) :
+    M →ₗ[R] N →ₗ[R] P := sorry
```
This feels like there is some categorical definition already, but I couldn't find it (without searching very hard).

Vierkantor commented on Mathlib/Algebra/Category/ModuleCat/Monoidal/Symmetric.lean:
```diff
@@ -29,6 +29,7 @@ namespace MonoidalCategory
 @[simp]
 theorem braiding_naturality {X₁ X₂ Y₁ Y₂ : ModuleCat.{u} R} (f : X₁ ⟶ Y₁) (g : X₂ ⟶ Y₂) :
     (f ⊗ g) ≫ (Y₁.braiding Y₂).hom = (X₁.braiding X₂).hom ≫ (g ⊗ f) := by
+  ext : 1
   apply TensorProduct.ext'
```
We should really fix the issue that makes `TensorProduct.ext'` not apply using the `ext` tactic...

Vierkantor commented on Mathlib/Algebra/Category/ModuleCat/Presheaf.lean:
```diff
@@ -267,7 +259,8 @@ noncomputable def restriction {X Y : Cᵒᵖ} (f : X ⟶ Y) :
   obj X := ModuleCat.of _ (R.obj X)
-  map {X Y } f :=
+  map {X Y} f := ModuleCat.asHom
+      (Y := (ModuleCat.restrictScalars (R.map f)).obj (ModuleCat.of (R.obj Y) (R.obj Y)))
```
More defeq issues.

Vierkantor commented on Mathlib/Algebra/Category/ModuleCat/Presheaf.lean:
```diff
@@ -388,6 +384,16 @@ noncomputable def forgetToPresheafModuleCatObj
     Cᵒᵖ ⥤ ModuleCat (R.obj X) where
   obj Y := forgetToPresheafModuleCatObjObj X hX M Y
   map f := forgetToPresheafModuleCatObjMap X hX M f
+  -- TODO: the lines below were automatic: probably there's some defeq mismatch between `M.obj` and
+  -- `forgetToPresheafModuleCatObjObj` in `forgetToPresheafModuleCatObjMap`...
+  map_id Y := by ext; erw [forgetToPresheafModuleCatObjMap_apply]; simp
+  map_comp f g := by
+    ext
+    simp only [ModuleCat.hom_comp, LinearMap.comp_apply]
+    erw [forgetToPresheafModuleCatObjMap_apply _ _ _ f,
+      forgetToPresheafModuleCatObjMap_apply _ _ _ g,
+      forgetToPresheafModuleCatObjMap_apply]
+    simp
```
More defeq issues...

Vierkantor commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -189,6 +192,12 @@ lemma smul_section_apply (r : R) (U : Opens (PrimeSpectrum.Top R))
 lemma smul_stalk_no_nonzero_divisor {x : PrimeSpectrum R}
     (r : x.asIdeal.primeCompl) (st : (tildeInModuleCat M).stalk x) (hst : r.1 • st = 0) :
     st = 0 := by
+  -- TODO: these instances are exactly the ones we wanted to avoid!
+  -- Probably best to change `colimit_no_zero_smul_divisor`
+  have : ∀ {c c' : ModuleCat R}, FunLike (c ⟶ c') ((forget (ModuleCat R)).obj c) ((forget (ModuleCat R)).obj c') :=
+    ConcreteCategory.instFunLike
+  have : ∀ {c c' : ModuleCat R}, LinearMapClass (c ⟶ c') R ((forget (ModuleCat R)).obj c) ((forget (ModuleCat R)).obj c') :=
+    sorry
```
This suggests `colimit_no_zero_smul_divisor` is not phrased correctly.

Vierkantor commented on Mathlib/RepresentationTheory/GroupCohomology/Basic.lean:
```diff
@@ -184,20 +184,22 @@ noncomputable abbrev inhomogeneousCochains : CochainComplex (ModuleCat k) ℕ :=
+      ((linearYonedaObjResolution A).d_comp_d n (n + 1) (n + 2)))
+    simp only [ModuleCat.hom_comp, LinearMap.comp_apply] at this
     dsimp only
-    simp only [d_eq, LinearEquiv.toModuleIso_inv, LinearEquiv.toModuleIso_hom, ModuleCat.coe_comp,
-      Function.comp_apply]
+    simp only [d_eq, LinearEquiv.toModuleIso_inv_hom, LinearEquiv.toModuleIso_hom_hom,
+      ModuleCat.hom_comp, LinearMap.comp_apply, LinearEquiv.coe_coe, ModuleCat.hom_zero]
     /- Porting note: I can see I need to rewrite `LinearEquiv.coe_coe` twice to at
       least reduce the need for `symm_apply_apply` to be an `erw`. However, even `erw` refuses to
       rewrite the second `coe_coe`... -/
-    erw [LinearEquiv.symm_apply_apply, this]
-    exact map_zero _
+    erw [LinearEquiv.symm_apply_apply, this, LinearMap.zero_apply]
+    simp
+    rfl
```
This proof got uglier, not sure what the concrete cause is.

Vierkantor replied:
I tried a few things and it seems there was already a lot of defeq abuse that got exposed now.

Vierkantor commented on Mathlib/RepresentationTheory/GroupCohomology/Basic.lean:
```diff
@@ -207,8 +209,15 @@ def inhomogeneousCochainsIso : inhomogeneousCochains A ≅ linearYonedaObjResolu
   subst h
-  simp only [CochainComplex.of_d, d_eq, Category.assoc, Iso.symm_hom, Iso.hom_inv_id,
-    Category.comp_id]
+  ext
+  simp only [ChainComplex.linearYonedaObj_X, linearYoneda_obj_obj_carrier, CochainComplex.of_x,
+    linearYoneda_obj_obj_isAddCommGroup, linearYoneda_obj_obj_isModule, Iso.symm_hom,
+    ChainComplex.linearYonedaObj_d, ModuleCat.hom_comp, linearYoneda_obj_map_hom,
+    Quiver.Hom.unop_op, LinearEquiv.toModuleIso_inv_hom, LinearMap.coe_comp, Function.comp_apply,
+    Linear.leftComp_apply, inhomogeneousCochains.d_def, d_eq, LinearEquiv.toModuleIso_hom_hom,
+    ModuleCat.asHom_comp, Category.assoc, LinearEquiv.comp_coe, LinearEquiv.self_trans_symm,
+    LinearEquiv.refl_toLinearMap, LinearMap.id_comp, LinearEquiv.coe_coe]
+  rfl
```
This proof got a lot uglier, not sure what the concrete cause is.

Vierkantor commented on Mathlib/RepresentationTheory/GroupCohomology/Resolution.lean:
```diff
@@ -70,6 +70,12 @@ universe u v w
 
 variable {k G : Type u} [CommRing k] {n : ℕ}
 
+-- TODO! Find a good home for me
+lemma ModuleCat.hom_sum {ι : Type*} {M N : ModuleCat k} (f : ι → (M ⟶ N)) (s : Finset ι) :
+    (∑ i in s, f i).hom = ∑ i in s, (f i).hom :=
+  map_sum ({ toFun := ModuleCat.Hom.hom, map_zero' := ModuleCat.hom_zero, map_add' :=
+    ModuleCat.hom_add } : (M ⟶ N) →+ (M →ₗ[k] N)) _ _
```
Should probably be moved to `ModuleCat/Basic.lean`, if that has the right imports.

Vierkantor commented on Mathlib/RepresentationTheory/Rep.lean:
```diff
@@ -57,15 +57,25 @@ instance (V : Rep k G) : Module k V := by
   change Module k ((forget₂ (Rep k G) (ModuleCat k)).obj V)
   infer_instance
 
+-- TODO: is there some abstract machinery for this?
+/-- `ModuleCat.Hom.hom` as an isomorphism of monoids. -/
+@[simps]
+def homMulEquiv (V : ModuleCat k) : End V ≃* (V →ₗ[k] V) where
+  toFun := ModuleCat.Hom.hom
+  invFun := ModuleCat.asHom
+  map_mul' _ _ := rfl
+  left_inv _ := rfl
+  right_inv _ := rfl
```
Is there a categorical name for this thing?

jcommelin replied:
It seems to me like this should work for endomorphism monoids in any `bundled` category.

Vierkantor committed:
Fix `AlgebraicGeometry/Modules/Tilde.lean` by specializing broken proofs

Vierkantor committed:
It builds!

Vierkantor committed:
Clean up proofs where we can drop the `erw`

Vierkantor committed:
Whitespace

Vierkantor committed:
Fix counterexamples

Vierkantor committed:
Cleanup

Vierkantor committed:
Fix instances in `ChangeOfRings`

Vierkantor committed:
Move definitions to their right homes

Vierkantor committed:
Clean up the redundant `asHomLeft` and `asHomRight` aliases

Vierkantor commented on Mathlib/CategoryTheory/Limits/ConcreteCategory/WithAlgebraicStructures.lean:
```diff
@@ -25,6 +24,11 @@ about `colimit F` when objects and morphisms in `C` have some algebraic structur
   and `x : Fᵢ` we have `r • x = 0` implies `x = 0`, then if `r • x = 0` for `x : colimit F`, then
   `x = 0`.
 
+## Implementation details
+
+For now, we specialize our results to `C = ModuleCat R`, which is the only place they are used.
+In the future they might be generalized by assuming a `HasForget₂ C (ModuleCat R)` instance,
+plus assertions that the module structures induced by `HasForget₂` coincide.
```
This is the major regression of this PR: we needed to specialize these results because there is no longer a good way to say for `ModuleCat` that all its homs are linear maps.

chrisflav commented on Mathlib/Algebra/Category/ModuleCat/Presheaf/Sheafify.lean:
```diff
@@ -314,10 +314,14 @@ def toSheafify : M₀ ⟶ (restrictScalars α).obj (sheafify α φ).val :=
-      ((toSheafify α φ).app X) x = φ.app X x := rfl
+    ((toSheafify α φ).app X).hom x = φ.app X x := rfl
+
+/-- `@[simp]`-normal form of `toSheafify_app_apply`. -/
+@[simp]
+lemma toSheafify_app_apply' (X : Cᵒᵖ) (x : M₀.obj X) :
+    DFunLike.coe (F := (_ →ₗ[_] ↑((ModuleCat.restrictScalars (α.app X)).obj _)))
+    ((toSheafify α φ).app X).hom x = φ.app X x := rfl
```
What's going on here?

chrisflav replied:
Ah I suppose this is related to the def-eq abuse you mention in the PR description concerning `ModuleCat.restrictScalars` right?

Vierkantor replied:
It's a bit more complicated but basically yes: we have that the codomain of `((toSheafify α φ).app X).hom` would by default be inferred as some object which is only definitionally equal to `(ModuleCat.restrictScalars (α.app X)).obj _))`. And these definitional equalities are `dsimp` lemmas so they do get applied before `toSheafify_app_apply` (which therefore would never get applied since it looks like it has the wrong type).

One idea I just had: maybe specify `no_index` instead of writing out the expected type tediously?

chrisflav replied:
Yes `no_index` works here

chrisflav commented on Mathlib/Algebra/Category/ModuleCat/Tannaka.lean:
```diff
@@ -39,19 +36,17 @@ def ringEquivEndForget₂ (R : Type u) [Ring R] :
     ext
-    dsimp
-    simp only [AddCommGrp.ofHom_apply, DistribMulAction.toAddMonoidHom_apply, add_smul]
+    simp [add_smul]
     rfl
```
I think `simp; rfl` is considered potentially unstable?

chrisflav commented on Mathlib/RepresentationTheory/GroupCohomology/Basic.lean:
```diff
@@ -184,20 +184,22 @@ noncomputable abbrev inhomogeneousCochains : CochainComplex (ModuleCat k) ℕ :=
       rewrite the second `coe_coe`... -/
-    erw [LinearEquiv.symm_apply_apply, this]
-    exact map_zero _
+    erw [LinearEquiv.symm_apply_apply, this, LinearMap.zero_apply]
+    simp
+    rfl
```
```suggestion
    erw [LinearEquiv.symm_apply_apply, this]
    simp only [LinearMap.zero_apply, ChainComplex.linearYonedaObj_X, linearYoneda_obj_obj_carrier,
      map_zero, Pi.zero_apply, LinearMap.zero_apply]
    rfl
```
maybe slightly better?

chrisflav committed:
fix some erws

chrisflav commented:
Overall this looks very good and a lot of proofs got cleaner! Did you run your magic `erw -> rw` tool? I could imagine there is some potential.

I hope you don't mind that I pushed some small changes to get rid of some more `erw`s.

chrisflav commented:
I tried to add a test file similar to the one for `AlgebraCat`, but I am running into this:
```
import Mathlib
open CategoryTheory ModuleCat
variable (R : Type u) [CommRing R]

/-
ambiguous, possible interpretations 
  ↟f : of R X ⟶ of R Y
  
  ↾⇑f : X ⟶ Y
-/
example {X Y : Type v} [AddCommGroup X] [Module R X] [AddCommGroup Y] [Module R Y] (f : X →ₗ[R] Y)
    (x : X) : (asHom f) x = f x := by simp
```
Do you understand this?

Vierkantor commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -88,27 +71,121 @@ instance : CoeSort (ModuleCat.{v} R) (Type v) :=
+  right f := ⟨⟨f⟩, rfl⟩
+
+/-- Typecheck a `LinearMap` as a morphism in `ModuleCat R`. -/
+abbrev asHom {X Y : Type v} [AddCommGroup X] [Module R X] [AddCommGroup Y] [Module R Y]
+    (f : X →ₗ[R] Y) : of R X ⟶ of R Y :=
+  ⟨f⟩
```
Yeah, I was very unsure which convention to follow: existing `ModuleCat.asHom` usage or consistency with the other categories. It's kind of funny that a couple months ago the deprecated alias was `ofHom` and now we want it to be exactly the opposite. :)

Vierkantor commented:
I think the testcase fails because there is also the definition [`CategoryTheory.asHom`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Types.html#CategoryTheory.asHom) which is `ModuleCat.asHom` but for the category of types. So we either apply `ModuleCat.asHom` and then coerce back to a function, or we coerce `f` to a function, apply `CategoryTheory.asHom` to make it a morphism, and then coerce it back to a plain function. And since both of the end results have the same types and applicability, Lean can't choose and throws an error.

chrisflav commented:
> I think the testcase fails because there is also the definition [`CategoryTheory.asHom`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Types.html#CategoryTheory.asHom) which is `ModuleCat.asHom` but for the category of types. So we either apply `ModuleCat.asHom` and then coerce back to a function, or we coerce `f` to a function, apply `CategoryTheory.asHom` to make it a morphism, and then coerce it back to a plain function. And since both of the end results have the same types and applicability, Lean can't choose and throws an error.

Ah, silly me, I didn't think of `CategoryTheory.asHom` and was not aware of the notation. Replacing the `asHom` with `ModuleCat.asHom` makes everything work.

chrisflav committed:
add tests for modulecat

chrisflav commented:
I put the tests in 554ba6315ef, if you like them feel free to cherry-pick (or tell me to push it here).

Vierkantor commented:
Thank you! Feel free to push your commit directly to this branch. :)

Vierkantor committed:
Define `ofHom` on `of`

Vierkantor committed:
Rename homMulEquiv to endMulEquiv

Vierkantor committed:
Replace `simp; rfl` with `simp only; rfl`

Vierkantor commented:
> Overall this looks very good and a lot of proofs got cleaner! Did you run your magic `erw -> rw` tool? I could imagine there is some potential.
> 
> I hope you don't mind that I pushed some small changes to get rid of some more `erw`s.

I did run it on an earlier version of the branch, let me see if anything got cleared up in the meantime.

(It should be as easy as merging the `try_this-erw` branch into the current branch, and then cherry-picking any of the warnings it produces.)

Vierkantor commented:
No free `erw` eliminations found on this branch, but I did find a few in other bits of Mathlib so time to make a little PR for that :D

chrisflav commented:
!bench

chrisflav commented:
The +227% one can be safely ignored, its a 60L file with one proof. Why do we not get the nice summary with all the instruction increases starting from 10^9? Does this mean, all of the changes are below that threshold? If yes, I think everything is fine.

Vierkantor commented:
`Mathlib.Algebra.Category.ModuleCat.Sheaf.Colimits` is also a one-proof file. If I can trust `set_option trace.profiler true` somewhat, then it looks like elaboration of that proof takes slightly less than `50%` longer, and kernel typechecking twice as long. But there is no obvious change in the steps taken to get there, just a bit of slowdown at every level...

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -88,27 +71,121 @@ instance : CoeSort (ModuleCat.{v} R) (Type v) :=
+
+@[deprecated (since := "2024-10-06")] alias ofHom := ModuleCat.asHom
+
+lemma hom_asHom {X Y : Type v} [AddCommGroup X] [Module R X] [AddCommGroup Y]
+    [Module R Y] (f : X →ₗ[R] Y) : (asHom f).hom = f := rfl
```
Should this also be `@[simp]`? If not, let's leave a comment in place explaining why.

Vierkantor replied:
It shouldn't be `@[simp]` because `asHom` and `hom` are reducibly defeq to the constructor and projection, respectively, so the whole thing reduces to `f = f`.

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -88,27 +71,121 @@ instance : CoeSort (ModuleCat.{v} R) (Type v) :=
 
-instance {M N : ModuleCat.{v} R} : LinearMapClass (M ⟶ N) R M N :=
-  LinearMap.semilinearMapClass
+lemma asHom_apply {M N : Type v} [AddCommGroup M] [AddCommGroup N] [Module R M] [Module R N]
+    (f : M →ₗ[R] N) (x : M) : asHom f x = f x := rfl
```
Similarly here, either `@[simp]` or explain why it's not a good idea.

Vierkantor replied:
I thought this would be the same issue, but actually it would be a good `@[simp]` lemma!

chrisflav replied:
This is solvable by `simp`, so I don't think we should add `@[simp]` here.

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -88,27 +71,121 @@ instance : CoeSort (ModuleCat.{v} R) (Type v) :=
+      map := fun f => f.hom }
+  forget_faithful := ⟨fun h => by ext x; simpa using congrFun h x⟩
+
+lemma forget_obj {M : ModuleCat.{v} R} : (forget (ModuleCat.{v} R)).obj M = M := rfl
+
+lemma forget_map {M N : ModuleCat.{v} R} (f : M ⟶ N) :
+    (forget (ModuleCat.{v} R)).map f = f :=
+  rfl
```
I think indeed these shouldn't be `simp`, but let's say so.

Vierkantor replied:
@chrisflav, am I correct that you removed these from the `simp` set in 6c554e33a35498ea9ef0fc288523c26d980e4fcc because:

* `forget_obj` will apply as a `dsimp` lemma to the (co)domain of morphisms, causing definitional equality issues?
* `forget_map` has a morphism on the LHS but a plain function on the RHS?

chrisflav replied:
But I also agree with your two points.

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -174,70 +235,37 @@ def ofSelfIso (M : ModuleCat R) : ModuleCat.of R M ≅ M where
+@[deprecated ModuleCat.asHom (since := "2024-11-29")] alias ModuleCat.asHomLeft := ModuleCat.asHom
 
 /-- Reinterpreting a linear map in the category of `R`-modules. -/
 scoped[ModuleCat] notation "↾" f:1024 => ModuleCat.asHomRight f
```
```suggestion
scoped[ModuleCat] notation "↾" f:1024 => ModuleCat.asHom f
```

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -174,70 +235,37 @@ def ofSelfIso (M : ModuleCat R) : ModuleCat.of R M ≅ M where
-/-- Reinterpreting a linear map in the category of `R`-modules. -/
-def ModuleCat.asHomLeft {X₁ : ModuleCat.{v} R} [AddCommGroup X₂] [Module R X₂] :
-    (X₁ →ₗ[R] X₂) → (X₁ ⟶ ModuleCat.of R X₂) :=
-  id
-
 /-- Reinterpreting a linear map in the category of `R`-modules. -/
 scoped[ModuleCat] notation "↿" f:1024 => ModuleCat.asHomLeft f
```
These are implicitly deprecated, right? Let's write something in the doc-string.

Vierkantor replied:
Indeed, since this notation resolves to the deprecated `asHomLeft`/`asHomRight`, we get a deprecation warning if we actually try to use the notation. I'll add the same warning to the docstring for the notation itself.

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -286,45 +317,116 @@ end
+instance : SMul S (M ⟶ N) where
+  smul c f := ⟨c • f.hom⟩
+
+@[simp] lemma hom_smul (n : S) (f : M ⟶ N) : (n • f).hom = n • f.hom := rfl
```
```suggestion
@[simp] lemma hom_smul (s : S) (f : M ⟶ N) : (s • f).hom = s • f.hom := rfl
```

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -286,45 +317,116 @@ end
+@[simp] lemma hom_zsmul (n : ℕ) (f : M ⟶ N) : (n • f).hom = n • f.hom := rfl
+
+instance : AddCommGroup (M ⟶ N) :=
+  Function.Injective.addCommGroup (Hom.hom) hom_bijective.injective
```
Should we just separately have `hom_injective`? Seems convenient.

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -286,45 +317,116 @@ end
+    hom_bijective.injective
+    (fun _ _ => rfl)
+
+/-- `ModuleCat.Hom.hom` bundled as a linear equivalence. -/
+def homLinearEquiv : (M ⟶ N) ≃ₗ[S] (M →ₗ[R] N) :=
+  { homEquiv with
+    map_add' := fun _ _ => rfl
+    map_smul' := fun _ _ => rfl }
```
This needs `simp` lemmas.

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -286,45 +317,116 @@ end
+    (fun _ _ => rfl)
+
+/-- `ModuleCat.Hom.hom` bundled as a linear equivalence. -/
+def homLinearEquiv : (M ⟶ N) ≃ₗ[S] (M →ₗ[R] N) :=
```
Should we separately have a `homAddEquiv`, before assuming the smul structure?

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -433,15 +535,40 @@ instance : (forget₂ (ModuleCat.{v} R) AddCommGrp.{v}).ReflectsIsomorphisms whe
+
+namespace ModuleCat
+
+/-- Turn a bilinear map into a homomorphism. -/
+@[simps]
+def asHom₂ {M N P : ModuleCat.{u} R} (f : M →ₗ[R] N →ₗ[R] P) :
+    M ⟶ of R (N ⟶ P) :=
+  asHom <| homLinearEquiv.symm.toLinearMap ∘ₗ f
```
Maybe this can just be a note for later, but I guess we could generalize this with an SMulCommClass, to `M : ModuleCat S`, `N P : ModuleCat R`.

Vierkantor replied:
I think that would make it difficult to make all the arrows typecheck since the objects would be in different categories.

(Also I am not sure myself which scalars you want to use for `(f : M →ₗ[_] N →ₗ[_] P)`.)

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -433,15 +535,40 @@ instance : (forget₂ (ModuleCat.{v} R) AddCommGrp.{v}).ReflectsIsomorphisms whe
+/-- Turn a homomorphism into a bilinear map. -/
+@[simps!]
+def Hom.hom₂ {M N P : ModuleCat.{u} R}
+    (f : ModuleCat.Hom M (of R (N ⟶ P))) :
```
Can we not write this with the categorical arrow?

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Basic.lean:
```diff
@@ -433,15 +535,40 @@ instance : (forget₂ (ModuleCat.{v} R) AddCommGrp.{v}).ReflectsIsomorphisms whe
+    M →ₗ[R] N →ₗ[R] P :=
+  Hom.hom (by convert (f ≫ asHom homLinearEquiv.toLinearMap))
+
+@[simp] lemma Hom.hom₂_asHom₂ {M N P : ModuleCat.{u} R} (f : M →ₗ[R] N →ₗ[R] P) :
+    (asHom₂ f).hom₂ = f := rfl
```
Other direction?

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Monoidal/Basic.lean:
```diff
@@ -95,63 +97,36 @@ instance instMonoidalCategoryStruct : MonoidalCategoryStruct (ModuleCat.{u} R) w
+  ext : 1
+  apply TensorProduct.ext_fourfold
+  intro w x y z
+  rfl
```
Nice!

kim-em commented on Mathlib/Algebra/Category/ModuleCat/Subobject.lean:
```diff
@@ -108,9 +106,7 @@ theorem cokernel_π_imageSubobject_ext {L M N : ModuleCat.{v} R} (f : L ⟶ M) [
-  change ((cokernel.π g) ∘ (g) ∘ (factorThruImageSubobject f)) l = 0
-  rw [← coe_comp, ← coe_comp, Category.assoc]
-  simp only [cokernel.condition, comp_zero]
-  rfl
+  change ((factorThruImageSubobject f) ≫ g ≫ (cokernel.π g)).hom l = 0
+  simp
```
The goal after `simp` here is 
```
(cokernel.π g).hom (g.hom ((factorThruImageSubobject f).hom l)) = 0
```
I think we should just have a simp lemma for `(cokernel.π g).hom (g.hom x) = 0`, called something like `ModuleCat.Hom.cokernel_condition`.

Ideally we'd have some tactic help, e.g. like `@[elementwise]`, but for now we should just write these at the point we set up cokernels.

We might also think (by introducing a level of indirection to `ModuleCat.Hom.hom`?) about how we could write a single simp lemma that would apply `cokernel_condition` to all of its different concrete versions.

Vierkantor replied:
I added a TODO comment as per our discussion.

kim-em commented on Mathlib/AlgebraicGeometry/Modules/Tilde.lean:
```diff
@@ -158,7 +158,10 @@ noncomputable instance (U : (Opens (PrimeSpectrum.Top R))ᵒᵖ) :
     { obj := fun U ↦ ModuleCat.of _ (M.tildeInType.val.obj U)
-      map := fun {U V} i ↦
+      map := fun {U V} i ↦ asHom
+        (X := of ((Spec (CommRingCat.of R)).ringCatSheaf.val.obj U) (M.tildeInType.val.obj U))
+        (Y := (restrictScalars ((Spec (CommRingCat.of R)).ringCatSheaf.val.map i)).obj
+          (of ((Spec (CommRingCat.of R)).ringCatSheaf.val.obj V) (M.tildeInType.val.obj V)))
```
Having to write out the source and target so often doesn't seem fun. :-(

At very least we should leave some comments on the places this is necessary, reminding us to come back.

I wonder if having a variant of `asHom` here, that took as input a semilinear map (twisted by `(Spec (CommRingCat.of R)).ringCatSheaf.val.map i`), could producing something that would typecheck more easily... Not sure yet.

kim-em replied:
Something like
```
def tildeInTypeLinearMap {U V : (Opens (PrimeSpectrum.Top R))ᵒᵖ} (i : U ⟶ V) :
    (M.tildeInType.1.obj U) →ₛₗ[(Spec (CommRingCat.of R)).ringCatSheaf.val.map i] 
        (of ((Spec (.of R)).ringCatSheaf.1.obj V) (M.tildeInType.1.obj V)) :=
  { toFun := M.tildeInType.val.map i
    map_smul' := by intros; rfl
    map_add' := by intros; rfl } 

def asHomRestrictScalars {M N : Type u} {R S : Type*} [Ring R] [Ring S] (σ : R →+* S)
    [AddCommGroup M] [Module R M] [AddCommGroup N] [Module S N] 
    (f : M →ₛₗ[σ] N) :
    ModuleCat.of R M ⟶ (restrictScalars σ).obj (ModuleCat.of S N) :=
  sorry

/--
`M^~` as a sheaf of `𝒪_{Spec R}`-modules
-/
noncomputable def tilde : (Spec (CommRingCat.of R)).Modules where
  val :=
    { obj := fun U ↦ ModuleCat.of _ (M.tildeInType.val.obj U)
      map := fun {U V} i ↦ 
        asHomRestrictScalars ((Spec (CommRingCat.of R)).ringCatSheaf.val.map i) (tildeInTypeLinearMap M i)
      map_id := sorry
      map_comp := sorry }
  isSheaf := (TopCat.Presheaf.isSheaf_iff_isSheaf_comp (forget AddCommGrp) _ ).2
    M.tildeInType.2
```

kim-em replied:
Why is `restrictScalars` printing with a `coe` arrow?? That is really unhelpful. :-)

Vierkantor replied:
I went through and added a comment at each point we have to add a type hint: looks like `restrictScalars` is always involved somehow.

Vierkantor committed:
Explain why this is not `@[simp]`

Vierkantor committed:
Kim's suggestions

Vierkantor committed:
Fix after merge

Vierkantor committed:
Add TODO comment

Vierkantor committed:
Add comments for each time we have to add a type hint.

It seems to be only with `restrictScalars` or definitions using `restrictScalars` such as in `AlgebraicGeometry/Modules/Tilde.lean`

Vierkantor committed:
Remove porting note on inserting `asHom` since this is now the design principle

Vierkantor committed:
Rename `asHom` -> `ofHom`

Vierkantor committed:
Fix tests

Vierkantor committed:
Merge remote-tracking branch 'origin/master' into ModuleCat-Hom-structure

jcommelin commented:
Hmmm, we still don't get the helpful report... Don't know why not :cry:

Vierkantor commented:
I checked again and am happy with the PR, so let's get this merged!

bors merge
